clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/cluster/AzureClusterProvider.groovy

```diff
@@ -0,0 +1,233 @@
+
+    if (!azureApplication) {
+      return [] as Set
+    }
+
+    def clusterKeys = []
+    azureApplication.clusterNames.each { String accountName, Set<String> clusterNames ->
+      clusterNames.each { String clusterName ->
+        clusterKeys << Keys.getClusterKey(azureCloudProvider, azureApplication.name, clusterName, accountName)
+      }
+    }
+
+    def clusters = cacheView.getAll(AZURE_CLUSTERS.ns, clusterKeys)
+
+    translateClusters(clusters, true) as Set<AzureCluster>
+  }
+
+  @Override
+  Map<String, Set<AzureCluster>> getClusters() {
+    Collection<CacheData> clusterData = cacheView.getAll(AZURE_CLUSTERS.ns)
+    Collection<AzureCluster> clusters = translateClusters(clusterData, true)
+    mapResponse(clusters)
+
+  }
+
+  @Override
+  Map<String, Set<AzureCluster>> getClusterSummaries(String applicationName) {
+    getClusters(applicationName, false)
+  }
+
+  @Override
+  Map<String, Set<AzureCluster>> getClusterDetails(String applicationName) {
+    getClusters(applicationName, true)
+  }
+
+  Map<String, Set<AzureCluster>> getClusters(String applicationName, Boolean includeInstanceDetails) {
+    CacheData appData = cacheView.get(AZURE_APPLICATIONS.ns, Keys.getApplicationKey(azureCloudProvider, applicationName))
+    if (!appData) {
+      return [:] as Map
+    }
+
+    Collection<AzureCluster> clusters = translateClusters(resolveRelationshipData(appData, AZURE_CLUSTERS.ns), includeInstanceDetails)
+    mapResponse(clusters)
+  }
+
+  @Override
+  AzureCluster getCluster(String applicationName, String account, String name) {
+    CacheData cluster = cacheView.get(AZURE_CLUSTERS.ns, Keys.getClusterKey(azureCloudProvider, applicationName, name, account, ))
+    cluster ? translateClusters([cluster], true)[0] : null
+  }
+
+  @Override
+  ServerGroup getServerGroup(String account, String region, String name) {
+    null
+  }
+
+  private Collection<AzureCluster> translateClusters(Collection<CacheData> clusterData, boolean includeDetails) {
+    Map<String, AzureLoadBalancer> loadBalancers
+    Map<String, AzureServerGroupDescription> serverGroups
+
+    if (includeDetails) {
+      Collection<CacheData> allLoadBalancers = resolveRelationshipDataForCollection(clusterData, AZURE_LOAD_BALANCERS.ns)
+      Collection<CacheData> allServerGroups = resolveRelationshipDataForCollection(clusterData, AZURE_SERVER_GROUPS.ns, RelationshipCacheFilter.include(AZURE_INSTANCES.ns))
+
+      loadBalancers = translateLoadBalancers(allLoadBalancers)
+      serverGroups = translateServerGroups(allServerGroups)
+    }
+
+    Collection<AzureCluster> clusters = clusterData.collect { CacheData clusterDataEntry ->
+      Map<String, String> clusterKey = Keys.parse(azureCloudProvider, clusterDataEntry.id)
+
+      def cluster = new AzureCluster()
+      cluster.accountName = clusterKey.account
+      cluster.name = clusterKey.name
+      cluster.type = azureCloudProvider.id
+      if (includeDetails) {
```

Comment:
It looks like the two if (includeDetails) tests could be combined into a single test which calls out to a separate method for adding in the details.